### PR TITLE
add cinema halls repo and create cinema halls use cases'

### PR DESCRIPTION
**Pull request by RubyDiver**
Is it possible to remove initializer from each use case and make one available to everyone so that I don't write it every time?

### CODE REVIEW

**Comment by Sonickers** [comment]
Yes, check out this sample project that Wojtek has sent to us :)
https://github.com/monterail/mirror-consono-backend/blob/main/lib/use_case/base.rb#L9

**Comment by jandudulski** [comment]
> Is it possible to remove initializer from each use case and make one available to everyone so that I don't write it every time?

IMHO there is nothing wrong in this kind of repetition. It is something that I call "repeated by accident", not by design, and in such cases DRY doesn't apply. If you feel that you have too much code to repeat you can think about introducing a little of custom DSL or ready solution like [dry-initializer](https://dry-rb.org/gems/dry-initializer/3.0/):

```ruby
module CinemaHalls
  module UseCases
    class Create
      extend Dry::Initializer

      param :repository, default: proc { CinemaHalls::Repository.new }
    end      
  end
end
```

File: app/controllers/api/v1/cinema_halls_controller.rb [hunk 1]
```diff
@@ -3,52 +3,47 @@ class Api::V1::CinemaHallsController < ApplicationController
...
   def index
-    @cinema_halls = CinemaHall.all
-
-    render json: @cinema_halls
+    CinemaHalls::UseCases::Index.new.call
   end
 
   def show
-
+    CinemaHalls::UseCases::Show.new.call(id: params[:id])
```
Comment (Sonickers) [inline app/controllers/api/v1/cinema_halls_controller.rb]: In this sample project you can also see how to get around the `.new.call` thing. :) 

```
def self.call(params:)
      new(params: params).call
end
```

File: app/concepts/cinema_halls/repository.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+module CinemaHalls
+  class Repository
+    attr_reader :adapter
+
+    def initialize(adapter: CinemaHalls)
```
Comment (tobiaszwaszak) [inline app/concepts/cinema_halls/repository.rb]: `adapter: CinemaHall`

File: app/controllers/api/v1/cinema_halls_controller.rb [hunk 3]
```diff
@@ -3,52 +3,47 @@ class Api::V1::CinemaHallsController < ApplicationController
...
   def create
-    @cinema_hall = CinemaHall.create(cinema_hall_params)
-
-    respond_to do |format|
-      if @cinema_hall.save
-        format.json {render :show, status: :created, location: @cinema_hall}
-      else
-        format.json {render json: @cinema_hall.errors, status: :unprocessable_entity}
-      end
+    @cinema_hall = CinemaHall::UseCases::Create.new.call(params: cinema_hall_params)
```
Comment (tobiaszwaszak) [inline app/controllers/api/v1/cinema_halls_controller.rb]: CinemaHalls

File: app/controllers/api/v1/cinema_halls_controller.rb [hunk 4]
```diff
@@ -3,52 +3,47 @@ class Api::V1::CinemaHallsController < ApplicationController
...
-      render :edit
+      render json: @cinema_hall.errors, status: :unprocessable_entity
     end
   end
 
   #DELETE /api/v1/cinema_hall/:id
   def destroy
-    @cinema_hall.destroy
-    redirect_to root_url, notice: 'Cinema hall has been deleted.'
+    CinemaHall::UseCases::Destroy.new.call(params: cinema_hall_id)
```
Comment (tobiaszwaszak) [inline app/controllers/api/v1/cinema_halls_controller.rb]: CinemaHalls

File: app/controllers/api/v1/cinema_halls_controller.rb [hunk 5]
```diff
@@ -3,52 +3,47 @@ class Api::V1::CinemaHallsController < ApplicationController
...
   end
 
   private
+
   def set_cinema_hall
```
Comment (tobiaszwaszak) [inline app/controllers/api/v1/cinema_halls_controller.rb]: You don't use it

**Pull request merged**